### DISCUSSION TITLE
Should the new static abstract members not be enforced in abstract classes?

### DISCUSSION THREAD

**Discussion by SoftCircuits** [dotnet/csharplang#7730] [Language Ideas]
I would like to put off implementing `static abstract` members to my derived class, as shown in the following example.

	public interface IReport
	{
		static abstract string Name { get; }
		static abstract string Description { get; }
	}

	public abstract class Report : IReport // ERROR: Must implement Name and Description
	{
            // Additional functionality
	}

	public class TestReport : Report
	{
		public static string Name => "Name";
		public static string Description => "Description";
	}

Here, I want `TestReport` to define `Name` and `Description`. So I made `Report` abstract.

But I still get an error for `Report` because it didn't define `Name` and `Description`.

**Comment by KennethHoff** [comment]
Related to https://github.com/dotnet/csharplang/issues/5783 and I agree that it's annoying.

A (poor) workaround is this:

```diff
public interface IReport
{
	static abstract string Name { get; }
	static abstract string Description { get; }
}

-public abstract class Report : IReport
+public abstract class Report
{
        // Additional functionality
}

-public class TestReport : Report
+public class TestReport : Report, IReport
{
	public static string Name => "Name";
	public static string Description => "Description";
}
```

**Reply by SoftCircuits** [reply]
Thanks. Now I need to figure out if I can still implement some `IReport` members in `Report` if I do it this way. I'll give it a try.

**Reply by SoftCircuits** [reply]
Unfortunately, you cannot do the following. This still gives an error for `TestReport` not implementing `DoSomething()`. So, taking this approach, I cannot put any common functionality from `IReport` in `Report`.

``` cs
public interface IReport
{
    static abstract string Name { get; }
    static abstract string Description { get; }
    void DoSomething();
}

public abstract class Report
{
    void DoSomething() { }
}

public class TestReport : Report, IReport
{
    public static string Name => "Name";
    public static string Description => "Description";
}
```

**Reply by TahirAhmadov** [reply]
Your `Report.DoSomething` is not public, that's why.

https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEDMACWBAU2QDMIBjIvASQCUiAHAe2QQFgAoAby737xYAjADY8EYAGcEySgkFCADHgByEMNW54A5kQQBuPAF8+A4WInTZFecOUARIpIrIojBFGYw8W3QeOm/FgALHj2zADKzBoIABaw2gAUAJT6XCacXLjiUjJyggBMeAwsbFy8nAKC+CFhkdF68TBJyT4BnBlZNUUAKk4IJazyIMVMQ3C0g2U8gdUKYnaq6tQAvAB8eABEahqbaZVmNaIKDk4ubh5eeOtbjs6u7p4we+lcWQU45bNYBQXcGUYgA

**Reply by SoftCircuits** [reply]
Oh, right you are! Cool. Many thanks!

**Comment by HaloFour** [comment]
This has always been the case with interfaces and abstract classes.  The required members of the interface aren't _inherited_ as "abstract" by the abstract class, they need to be explicitly defined:

```cs
public interface I {
    void M();
}

public abstract class C : I { // error CS0535: 'C' does not implement interface member 'I.M()'
}
```
